datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java

```diff
@@ -96,4 +107,33 @@ private void validateThatEveryInstanceIsPresentInTheConfig(Map<String, List<Stri
                     String.format("The following instances are missing from the ordered OS upgrade request: %s", instancesFromInstanceGroups));
         }
     }
+
+    private void addServiceHostGroupsToOrderedOSUpgradeSet(Map<String, List<String>> instanceIdsByInstanceGroup, int order,
+            List<OrderedOSUpgradeSet> osUpgradeByUpgradeSets) {
+        Set<String> instanceIds = null;
+        do {
+            if (instanceIds != null) {
+                osUpgradeByUpgradeSets.add(new OrderedOSUpgradeSet(order++, instanceIds));
+            }
+            instanceIds = getServicesHostGroupIds(instanceIdsByInstanceGroup);
+        } while (!instanceIds.isEmpty());
+    }
+
+    private Set<String> getServicesHostGroupIds(Map<String, List<String>> instanceIdsByInstanceGroup) {
+        Set<String> instanceIds = new HashSet<>();
+        addInstanceId(instanceIdsByInstanceGroup, SOLRHG, instanceIds);
+        addInstanceId(instanceIdsByInstanceGroup, STORAGEHG, instanceIds);
+        addInstanceId(instanceIdsByInstanceGroup, KAFKAHG, instanceIds);
+        addInstanceId(instanceIdsByInstanceGroup, RAZHG, instanceIds);
+        addInstanceId(instanceIdsByInstanceGroup, ATLASHG, instanceIds);
+        addInstanceId(instanceIdsByInstanceGroup, HMSHG, instanceIds);
+        return instanceIds;
+    }
+
+    private void addInstanceId(Map<String, List<String>> instanceIdsByInstanceGroup, InstanceGroupName instanceGroupName, Set<String> instanceIds) {
```

Comment:
Please rename it to something more meaningful like pollInstanceIdFromServiceHostGroup